add signals and queries design patterns (#733)
temporalio/documentation: docs/typescript/workflows.md

#### Why not `new Signal` and `new Query`?

The semantic of `defineSignal`/`defineQuery` is intentional, in that they return Signal/Query **Definitions**, not unique instances of Signals and Queries themselves. [View source](https://github.com/temporalio/sdk-typescript/blob/fc658d3760e6653aec47732ab17a0062b7dd23fc/packages/workflow/src/workflow.ts#L884-L907):

```ts
/**
 * Define a signal method for a Workflow.
 */
export function defineSignal<Args extends any[] = []>(
  name: string
): SignalDefinition<Args> {
  return {
    type: 'signal',
    name,
  };
}

/**
 * Define a query method for a Workflow.
 */
export function defineQuery<Ret, Args extends any[] = []>(
  name: string
): QueryDefinition<Ret, Args> {
  return {
    type: 'query',
    name,
  };
}
```

Signals/Queries are only instantiated in `setHandler` and are specific to a particular Workflow Execution.

These distinctions may seem minor, but they model how Temporal works under the hood, because Signals and Queries are messages identified by "just strings" and don't have meaning independent of the Workflow having a listener to handle them.
This will be clearer if you refer to to the Client-side APIs below.

> #### Why `setHandler` and not OTHER_API?
> 
> We named it `setHandler` instead of `subscribe` because Signals/Queries can only have one listener at a time, whereas `subscribe` could imply an Observable with multiple consumers.

Author: ```suggestion
#### Why `setHandler` and not OTHER_API?

We named it `setHandler` instead of `subscribe` because Signals/Queries can only have one "handler" at a time, whereas `subscribe` could imply an Observable with multiple consumers, and is a higher level construct.
```

#### Why not `new Signal` and `new Query`?

The semantic of `defineSignal`/`defineQuery` is intentional, in that they return Signal/Query **Definitions**, not unique instances of Signals and Queries themselves. [View source](https://github.com/temporalio/sdk-typescript/blob/fc658d3760e6653aec47732ab17a0062b7dd23fc/packages/workflow/src/workflow.ts#L884-L907):
  
```ts
/**
 * Define a signal method for a Workflow.
 *

>  * Definitions are used to register handler in the Workflow via {@link setHandler} and to signal Workflows using a {@link WorkflowHandle}, {@link ChildWorkflowHandle} or {@link ExternalWorkflowHandle}.
>  * Definitions can be reused in multiple Workflows.
>  */
> export function defineSignal<Args extends any[] = []>(name: string): SignalDefinition<Args> {
>   return {
>     type: 'signal',
>     name,
>   };
> }
> 
> /**
>  * Define a query method for a Workflow.
>  *
>  * Definitions are used to register handler in the Workflow via {@link setHandler} and to query Workflows using a {@link WorkflowHandle}.
>  * Definitions can be reused in multiple Workflows.

Author: ```suggestion
 */
export function defineSignal<Args extends any[] = []>(name: string): SignalDefinition<Args> {
  return {
    type: 'signal',
    name,
  };
}

/**
 * Define a query method for a Workflow.
 *
```

#### Why not `new Signal` and `new Query`?

The semantic of `defineSignal`/`defineQuery` is intentional, in that they return Signal/Query **Definitions**, not unique instances of Signals and Queries themselves. [View source](https://github.com/temporalio/sdk-typescript/blob/fc658d3760e6653aec47732ab17a0062b7dd23fc/packages/workflow/src/workflow.ts#L884-L907):

```ts

> /**
>  * Define a signal method for a Workflow.
>  *
>  * Definitions are used to register handler in the Workflow via {@link setHandler} and to signal Workflows using a {@link WorkflowHandle}, {@link ChildWorkflowHandle} or {@link ExternalWorkflowHandle}.
>  * Definitions can be reused in multiple Workflows.
>  */
> export function defineSignal<Args extends any[] = []>(
>   name: string
> ): SignalDefinition<Args> {
>   return {
>     type: 'signal',
>     name,
>   };
> }
> 
> /**
>  * Define a query method for a Workflow.
>  *
>  * Definitions are used to register handler in the Workflow via {@link setHandler} and to query Workflows using a {@link WorkflowHandle}.
>  * Definitions can be reused in multiple Workflows.
>  */
> export function defineQuery<Ret, Args extends any[] = []>(
>   name: string
> ): QueryDefinition<Ret, Args> {
>   return {
>     type: 'query',
>     name,
>   };
> }

Author: ```suggestion
/**
 * Define a signal method for a Workflow.
 */
export function defineSignal<Args extends any[] = []>(name: string): SignalDefinition<Args> {
  return {
    type: 'signal',
    name,
  };
}

/**
 * Define a query method for a Workflow.
 */
export function defineQuery<Ret, Args extends any[] = []>( name: string): QueryDefinition<Ret, Args> {
  return {
    type: 'query',
    name,
  };
}
```
